### PR TITLE
Add selinux policy per distro/version

### LINKED ISSUE #25217: Error when installing on F24
**Output of `docker version`:**

```
Client:
 Version:      1.12.0
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   8eab29e
 Built:        
 OS/Arch:      linux/amd64

Server:
 Version:      1.12.0
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   8eab29e
 Built:        
 OS/Arch:      linux/amd64
```

**Output of `docker info`:**

```
Containers: 7
 Running: 0
 Paused: 0
 Stopped: 7
Images: 219
Server Version: 1.12.0
Storage Driver: devicemapper
 Pool Name: docker-253:0-4981931-pool
 Pool Blocksize: 65.54 kB
 Base Device Size: 10.74 GB
 Backing Filesystem: ext4
 Data file: /dev/loop0
 Metadata file: /dev/loop1
 Data Space Used: 4.523 GB
 Data Space Total: 107.4 GB
 Data Space Available: 50.5 GB
 Metadata Space Used: 11.76 MB
 Metadata Space Total: 2.147 GB
 Metadata Space Available: 2.136 GB
 Thin Pool Minimum Free Space: 10.74 GB
 Udev Sync Supported: true
 Deferred Removal Enabled: false
 Deferred Deletion Enabled: false
 Deferred Deleted Device Count: 0
 Data loop file: /var/lib/docker/devicemapper/devicemapper/data
 WARNING: Usage of loopback devices is strongly discouraged for production use. Use `--storage-opt dm.thinpooldev` to specify a custom block storage device.
 Metadata loop file: /var/lib/docker/devicemapper/devicemapper/metadata
 Library Version: 1.02.122 (2016-04-09)
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge null host overlay
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Security Options: seccomp
Kernel Version: 4.6.4-301.fc24.x86_64
Operating System: Fedora 24 (Workstation Edition)
OSType: linux
Architecture: x86_64
CPUs: 8
Total Memory: 15.65 GiB
Name: aschoeneck.united.domain
ID: SLJS:ITNI:R22Q:WKKD:ELUT:LGPV:CDSY:LF7M:3JTY:YT3D:SLE3:EYLU
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Insecure Registries:
 127.0.0.0/8
```

**Additional environment details**
physical

**Steps to reproduce the issue:**
1. `dnf install docker-engine`

**Describe the results you received:**
```
Transaktion wird ausgeführt
  Installieren     : docker-engine-selinux-1.12.0-1.fc24.noarch                                                                                             1/2 
Re-declaration of boolean virt_sandbox_use_fusefs
Failed to create node
Bad boolean declaration at /var/lib/selinux/targeted/tmp/modules/100/virt/cil:148
/usr/sbin/semodule:  Failed!
```

**Describe the results you expected:**
Installation succeeds without any errors

### PR DESCRIPTION
**Pull request by crosbymichael**
This adds the ability to have different profiles for individual distros
and versions of the distro because they all ship with and depend on
different versions of policy packages.

The `selinux` dir contains the unmodified policy that is being used
today.  The `selinux-fedora` dir contains the new policy for fedora 24
with the changes for it to compile and work on the system.

The fedora policy is from commit
https://github.com/projectatomic/docker-selinux/commit/4a6ce94da5e34868642ebcc7a67da937d4c73283

Fixes #25217

Signed-off-by: Michael Crosby crosbymichael@gmail.com

### CODE REVIEW

**Comment by crosbymichael** [comment]
This also provides the "framework" so that oraclelinux can have their own custom policy that is inline with what they ship to resolve #24612.  They will open a followup PR with a policy based on the same fedora one with a new omissions.

**Comment by crosbymichael** [comment]
ping @tiborvass @mlaventure

File: hack/make/build-rpm [hunk 1]
```diff
@@ -135,8 +135,14 @@ set -e
 		# selinux policy referencing systemd things won't work on non-systemd versions
 		# of centos or rhel, which we don't support anyways
 		if [ "${suite%.*}" -gt 6 ] && [[ "$version" != opensuse* ]]; then
+            selinuxDir="selinux"
```
Comment (mlaventure) [inline hack/make/build-rpm]: alignment issue

File: hack/make/build-rpm [hunk 2]
```diff
@@ -135,8 +135,14 @@ set -e
...
+            case "$version" in
+                fedora-24)
+                    selinuxDir="selinux-fedora"
```
Comment (mlaventure) [inline hack/make/build-rpm]: can the folder just be called `selinux-fedora-24` that way you can just test if it exists and replace the default value with it in that case ? (and it'll be faster to add new exceptions)

Comment (mlaventure) [inline hack/make/build-rpm]: or the reverse:

```
selinuxDir="selinux-${version}"`

[ ! -d ${selinuxDir} ] && selinuxDir="selinux"
```

Comment (crosbymichael) [inline hack/make/build-rpm]: you highly overestimate my bash skillz

**Comment by jimmidyson** [comment]
Tried this on Fedora 24 & didn't solve #24612 for me. Built RPM with `make rpm` before trying to upgrade. Result below.

```
$ sudo dnf install ./docker-engine-selinux-1.13.0-0.0.20160801.230201.gite781042.fc24.noarch.rpm
Dependencies resolved.
=================================================================================================================
 Package                    Arch        Version                                          Repository         Size
=================================================================================================================
Upgrading:
 docker-engine-selinux      noarch      1.13.0-0.0.20160801.230201.gite781042.fc24       @commandline       31 k

Transaction Summary
=================================================================================================================
Upgrade  1 Package

Total size: 31 k
Is this ok [y/N]: y
Downloading Packages:
Running transaction check
Transaction check succeeded.
Running transaction test
Transaction test succeeded.
Running transaction
  Upgrading   : docker-engine-selinux-1.13.0-0.0.20160801.230201.gite781042.fc24.noarch                      1/2 
Re-declaration of boolean virt_sandbox_use_fusefs
Failed to create node
Bad boolean declaration at /var/lib/selinux/targeted/tmp/modules/100/virt/cil:148
/usr/sbin/semodule:  Failed!
  Cleanup     : docker-engine-selinux-1.12.0-1.fc24.noarch                                                   2/2 
  Verifying   : docker-engine-selinux-1.13.0-0.0.20160801.230201.gite781042.fc24.noarch                      1/2 
  Verifying   : docker-engine-selinux-1.12.0-1.fc24.noarch                                                   2/2 

Upgraded:
  docker-engine-selinux.noarch 1.13.0-0.0.20160801.230201.gite781042.fc24                                        

Complete!
```

**Comment by crosbymichael** [comment]
@jimmidyson it should be fixed now.  I had a bug in my bash abilities and it was not building from the right dir on fedora

**Comment by jimmidyson** [comment]
Install now succeeds but permission problem still exists:

```
$ dnf install docker-engine-selinux-1.13.0-0.0.20160801.230201.git32b1f26.fc24.noarch.rpm        
Last metadata expiration check: 0:03:08 ago on Thu Aug  4 23:58:50 2016.
Dependencies resolved.
=================================================================================================================
 Package                     Arch        Version                                         Repository         Size
=================================================================================================================
Installing:
 docker-engine-selinux       noarch      1.13.0-0.0.20160801.230201.git32b1f26.fc24      @commandline       32 k
 policycoreutils-python      x86_64      2.5-13.fc24                                     updates           401 k

Transaction Summary
=================================================================================================================
Install  2 Packages

Total size: 433 k
Total download size: 401 k
Installed size: 1.1 M
Is this ok [y/N]: y
Downloading Packages:
policycoreutils-python-2.5-13.fc24.x86_64.rpm                                    2.0 MB/s | 401 kB     00:00    
-----------------------------------------------------------------------------------------------------------------
Total                                                                            319 kB/s | 401 kB     00:01     
Running transaction check
Transaction check succeeded.
Running transaction test
Transaction test succeeded.
Running transaction
  Installing  : policycoreutils-python-2.5-13.fc24.x86_64                                                    1/2 
  Installing  : docker-engine-selinux-1.13.0-0.0.20160801.230201.git32b1f26.fc24.noarch                      2/2 
  Verifying   : docker-engine-selinux-1.13.0-0.0.20160801.230201.git32b1f26.fc24.noarch                      1/2 
  Verifying   : policycoreutils-python-2.5-13.fc24.x86_64                                                    2/2 

Installed:
  docker-engine-selinux.noarch 1.13.0-0.0.20160801.230201.git32b1f26.fc24                                        
  policycoreutils-python.x86_64 2.5-13.fc24                                                                      

Complete!

$ docker run -it --rm busybox                        
Unable to find image 'busybox:latest' locally
latest: Pulling from library/busybox
8ddc19f16526: Pull complete 
Digest: sha256:a59906e33509d14c036c8678d687bd4eec81ed7c4b8ce907b888c607f6a1e0e6
Status: Downloaded newer image for busybox:latest
standard_init_linux.go:175: exec user process caused "permission denied"
```

**Comment by crosbymichael** [comment]
if they updated the package on fedora again....

**Comment by jimmidyson** [comment]
Sorry I'm away until Monday now so won't be able to test out any updates. Should have checked audit log before I switched off...

**Comment by crosbymichael** [comment]
I tried this on a fresh vm again and it works, no perm error.

**Comment by jimmidyson** [comment]
I've just cleaned up & reinstalled as well & all looks good. Not sure what went wrong before, perhaps it was because docker daemon wasn't restarted after appling selinux policy? If binary gets relabelled does the daemon need restarting before context is applied? If so, should installing new version of docker selinux RPM cause docker daemon to be restarted to run under the new context?

**Comment by crosbymichael** [comment]
@jimmidyson not sure.  i'll have to test the upgrade path to see if that is the case

**Comment by jimmidyson** [comment]
Yeah needs a restart, unless there's some other way to make it the runtime context:

```
$ rpm -q docker-engine-selinux           
docker-engine-selinux-1.12.0-1.fc24.noarch

$  docker run -it --rm busybox
standard_init_linux.go:175: exec user process caused "permission denied"

$ dnf install ./docker-engine-selinux-1.13.0-0.0.20160801.230201.git32b1f26.fc24.noarch.rpm
Last metadata expiration check: 0:07:51 ago on Tue Aug  9 08:44:35 2016.
Dependencies resolved.
==============================================================================================
 Package               Arch   Version                                      Repository    Size
==============================================================================================
Upgrading:
 docker-engine-selinux noarch 1.13.0-0.0.20160801.230201.git32b1f26.fc24   @commandline  32 k

Transaction Summary
==============================================================================================
Upgrade  1 Package

Total size: 32 k
Is this ok [y/N]: y
Downloading Packages:
Running transaction check
Transaction check succeeded.
Running transaction test
Transaction test succeeded.
Running transaction
  Upgrading   : docker-engine-selinux-1.13.0-0.0.20160801.230201.git32b1f26.fc24.noarch   1/2 
  Cleanup     : docker-engine-selinux-1.12.0-1.fc24.noarch                                2/2 
  Verifying   : docker-engine-selinux-1.13.0-0.0.20160801.230201.git32b1f26.fc24.noarch   1/2 
  Verifying   : docker-engine-selinux-1.12.0-1.fc24.noarch                                2/2 

Upgraded:
  docker-engine-selinux.noarch 1.13.0-0.0.20160801.230201.git32b1f26.fc24                     

Complete!

$ docker run -it --rm busybox
standard_init_linux.go:175: exec user process caused "permission denied"

$ systemctl restart docker

$ docker run -it --rm busybox
/ #
```

**Comment by tiborvass** [comment]
LGTM

**Comment by jimmidyson** [comment]
Did the required restart get added?

**Comment by jimmidyson** [comment]
@tiborvass @crosbymichael So did the required restart get added to update script?

**Comment by crosbymichael** [comment]
@jimmidyson because we ship the profile with the docker package it will take care of the restart.  We will be shipping with with 1.12.1 along with a few other bug fixes

**Comment by jimmidyson** [comment]
Ah OK so not upgradable independently? Fair enough.

**Comment by crosbymichael** [comment]
@jimmidyson not right now.  I would like to decouple this in the future so that we can push updates independently but right now they are released in lockstep.

**Pull request merged**